src/python.cpp

```diff
@@ -1223,15 +1307,33 @@ PyObject* r_to_py_cpp(RObject x, bool convert) {
 
   // numeric (pass length 1 vectors as scalars, otherwise pass list)
   } else if (type == REALSXP) {
+    bool isBit64  = x.inherits("integer64");
+    bool isUnsigned = x.inherits("np.ulong");
+    bool isInt64 = isBit64 && !isUnsigned;
+    bool isUint64 = isBit64 && isUnsigned;
     if (LENGTH(sexp) == 1) {
       double value = REAL(sexp)[0];
-      return PyFloat_FromDouble(value);
+      if(isInt64) {
+        return PyInt_FromLong(reinterpret_cast<long&>(value));
```

Comment:
Is it really safe to reinterpret a value as a reference? I've usually seen type punning done as e.g.

```
*(long*)(&value)
```

This is still technically undefined behavior, though. My understanding is that the only standards-compliant way of performing type punning is with `memcpy()`. See: https://stackoverflow.com/a/17790026/1342082

(In C++20, we will have bitcast: https://en.cppreference.com/w/cpp/numeric/bit_cast)

Reply:
See https://godbolt.org/z/VDCcVF for an example of a helper function that could be used for casting here.